src/target/fhe/trans.rs

```diff
@@ -0,0 +1,245 @@
+
+                match n.len() {
+                    5 => n[3].to_string(),
+                    6.. => {
+                        let l = n.len() - 1;
+                        format!("{}_{}", n[l - 2], n[l])
+                    }
+                    _ => {
+                        panic!("Invalid variable name: {}", name);
+                    }
+                }
+            }
+            _ => panic!("Term {} is not of type Var", t),
+        }
+    }
+
+    fn unwrap_vis(&self, name: &str) -> u8 {
+        match self.md.input_vis.get(name).unwrap() {
+            Some(_) => 1,
+            None => 0,
+        }
+    }
+
+    fn embed_eq(&mut self, t: Term, a_term: Term, b_term: Term) {
+        let s = self.term_to_stext_cnt.get(&t).unwrap();
+        let a = self.term_to_stext_cnt.get(&t.cs[0]).unwrap();
+        let b = self.term_to_stext_cnt.get(&t.cs[1]).unwrap();
+        let op = "EQ";
+        let line = format!("2 1 {} {} {} {}\n", a, b, s, op);
+        self.bytecode_output.push(line);
+        match check(&a_term) {
+            Sort::Bool => {
+                self.check_bool(&a_term);
+                self.check_bool(&b_term);
+                self.cache.insert(t, EmbeddedTerm::Bool);
+            }
+            Sort::BitVector(_) => {
+                self.check_bv(&a_term);
+                self.check_bv(&b_term);
+                self.cache.insert(t, EmbeddedTerm::Bool);
+            }
+            e => panic!("Unimplemented sort for Eq: {:?}", e),
+        }
+    }
+
+    /// Given term `t`, type-check `t` is of type Bool
+    fn check_bool(&self, t: &Term) {
+        self.cache
+            .get(t)
+            .unwrap_or_else(|| panic!("Missing expression for {:?}", t));
+    }
+
+    fn embed_bool(&mut self, t: Term) {
+        let s = self.term_to_stext_cnt.get(&t).unwrap();
+        match &t.op {
+            Op::Var(name, Sort::Bool) => {
+                // write to bytecode file
+                if !self.inputs.contains(&t) && self.md.input_vis.contains_key(name) {
+                    let term_name = ToFHE::get_var_name(&t);
+                    let enc = self.unwrap_vis(name);
+                    let share_cnt = self.term_to_stext_cnt.get(&t).unwrap();
+                    let op = "IN";
+
+                    let line = format!("2 1 {} {} {} {}\n", term_name, enc, share_cnt, op);
+                    self.bytecode_output.insert(0, line);
+                }
+                if !self.cache.contains_key(&t) {
+                    self.cache.insert(t.clone(), EmbeddedTerm::Bool);
+                }
+            }
+            Op::Const(Value::Bool(b)) => {
+                let op = "CONS_bool";
+                let line = format!("1 1 {} {} {}\n", *b as i32, s, op);
+                self.bytecode_output.push(line);
+                self.cache.insert(t.clone(), EmbeddedTerm::Bool);
+            }
+            Op::Eq => {
+                self.embed_eq(t.clone(), t.cs[0].clone(), t.cs[1].clone());
```

Comment:
If this is implemented, then we need to add a test for it.

Reply:
I haven't implemented it on the interpreter side yet